atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/DialogueClientFactory.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.http.v2;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+import org.immutables.value.Value;
+
+import com.google.common.util.concurrent.Futures;
+import com.palantir.atlasdb.config.ServerListConfig;
+import com.palantir.conjure.java.api.config.service.UserAgent;
+import com.palantir.conjure.java.client.config.ClientConfiguration;
+import com.palantir.conjure.java.client.jaxrs.JaxRsClient;
+import com.palantir.conjure.java.dialogue.serde.DefaultConjureRuntime;
+import com.palantir.dialogue.Channel;
+import com.palantir.dialogue.ConjureRuntime;
+import com.palantir.dialogue.core.DialogueChannel;
+import com.palantir.dialogue.hc4.ApacheHttpClientChannels;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
+import com.palantir.logsafe.exceptions.SafeIllegalStateException;
+import com.palantir.refreshable.Refreshable;
+import com.palantir.tritium.Tritium;
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
+
+final class DialogueClientFactory {
+    private static final DefaultConjureRuntime DIALOGUE_RUNTIME =
+            DefaultConjureRuntime.builder().build();
+    private static final String SERVICE_NAME = "lock";
+    private final TaggedMetricRegistry taggedMetricRegistry;
+    private final Refreshable<Optional<ServerListConfig>> config;
+    private final UserAgent userAgent;
+    private volatile ConcurrentHashMap<CacheKey, DialogueChannel> cachedChannels = new ConcurrentHashMap<>();
```

Comment:
Why is this volatile?

Reply:
Why do we have this map in the first place? Is this so that you can reuse the channels if we have, say, multiple services on the same host we talk to (so ConjureTimelockService and something else in the future?)

Reply:
We need to make sure that this is closed whenever TransactionManager is closed.